utils/src/main/java/org/killbill/commons/utils/cache/DefaultCache.java

```diff
@@ -0,0 +1,133 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.killbill.commons.utils.cache;
+
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.function.Function;
+
+import org.killbill.commons.utils.annotation.VisibleForTesting;
+
+/**
+ * <p>
+ *     Default {@link Cache} implementation, that provide ability to:
+ *     <ol>
+ *         <li>Add maxSize to the cache. If more entry added, the oldest entry get removed automatically</li>
+ *         <li>Add lazy-loading capability with {@code cacheLoader} parameter in constructor. This {@code cacheLoader}
+ *             will be called if {@link #get(Object)} or {@link #getOrLoad(Object, Function)} return {@code null}</li>
+ *         <li>Add timout (similar to expire-after-write in Guava and Caffeine) capability</li>
+ *     </ol>
+ * </p>
+ * @param <K> cache key
+ * @param <V> cache value
+ */
+public class DefaultCache<K, V> implements Cache<K, V> {
+
+    public static final long NO_TIMEOUT = 0;
+
+    @VisibleForTesting
+    final Map<K, TimedValue<V>> map;
+
+    /** Timeout value in second */
+    private final long timeout;
+    private final Function<K, V> cacheLoader;
+
+    /**
+     * Create cache with maximum entry size, without any timout (live forever) and no cache loader.
+     *
+     * @param maxSize max entry that should be existed in cache.
+     */
+    public DefaultCache(final int maxSize) {
+        this(maxSize, NO_TIMEOUT, noCacheLoader());
+    }
+
+    /**
+     * Create cache with maximum entry size, timeout (in second), and cacheLoader capability.
+     *
+     * @param maxSize cache maximum size. If more entry added, the oldest entry get removed automatically.
+     * @param timeout
+     * @param cacheLoader
+     */
+    public DefaultCache(final int maxSize, final long timeout, final Function<K, V> cacheLoader) {
+        this.timeout = timeout * 1_000;
+        this.cacheLoader = cacheLoader;
+
+        map = Collections.synchronizedMap(new LinkedHashMap<>(16, 0.75f, true) {
+            @Override
+            protected boolean removeEldestEntry(final Entry<K, TimedValue<V>> eldest) {
+                return size() > maxSize;
+            }
+        });
+    }
+
+    public static <K1, V1> Function<K1, V1> noCacheLoader() {
+        return k1 -> null;
+    }
+
+    protected boolean isTimeoutEnabled() {
+        return timeout > 0L;
+    }
+
+    protected boolean isCacheLoaderExist() {
+        return !noCacheLoader().equals(cacheLoader);
+    }
+
+    protected void evictExpireEntry() {
+        if (isTimeoutEnabled()) {
+            map.values().removeIf(TimedValue::isTimeout);
```

Comment:
This could be quite costly for large maps. How about evicting a specific entry instead?

See below:

```
final TimedValue<V> timedValue = map.get(key);
if (timedValue.isTimeout) {
  invalidate(key);
}
// Reload
```